src/abi/json-abi-format.md

```diff
@@ -35,52 +36,39 @@ The ABI of a contract is represented as a JSON object containing the following p
+  - `"concreteTypeId"`: a unique string hash based ID. Generated as specified in [Hash Based Ids](./hash-based-ids.md).
+  - `"metadataTypeId"`: the the _type metadata declaration_ ID of this type, if the type metadata has components or is generic, otherwise non existent.
+  - `"typeArguments"`: an array of _type concrete declarations_ hash based IDs of the type parameters of the type, if the type is generic, otherwise non existent.
+- `"typesMetadata"`: an array describing all the _type metadata declarations_ used (or transitively used) in the ABI. Each _type metadata declaration_ is a JSON object that contains the following properties:
+  - `"type"`: a string representation of the _type metadata declaration_. The section [JSON ABI Format for Each Possible Metadata Type Declaration](#json-abi-format-for-each-possible-metadata-type-declaration) specifies the format for each possible type.
+  - `"metadataTypeId"`: a unique integer ID.
+  - `"components"`: an array of the components of a given type, if any, otherwise non existent. Each component is a _type application_ represented as a JSON object that contains the following properties:
     - `"name"`: the name of the component.
-    - `"type"`: the _type declaration_ hash based ID of the type of the component.
-    - `"typeArguments"`: an array of the _type arguments_ used when applying the type of the component, if the type is generic, and `null` otherwise. Each _type argument_ is a _type application_ represented as a JSON object that contains the following properties:
-      - `"type"`: the _type declaration_ hash based ID of the type of the _type argument_.
-      - `"typeArguments"`: an array of the _type arguments_ used when applying the type of the _type argument_, if the type is generic, and `null` otherwise. The format of the elements of this array recursively follows the rules described in this section.
-  - `"typeParameters"`: an array of type hash based IDs of the _type parameters_ of the type, if the type is generic, and `null` otherwise. Each _type parameter_ is a type declaration and is represented as described in [Generic Type Parameter](#generic-type-parameter).
+    - `"typeId"`: the _type metadata declaration_ ID (number) or _type concrete declaration_ hash based ID (string) of the type of the component.
+    - `"typeArguments"`: an array of the _type arguments_ used when applying the type of the component, if the type is generic, otherwise non existent. Each _type argument_ is a _type application_ represented as a JSON object that contains the following properties:
+      - `"typeId"`: the _type metadata declaration_ ID (number) or _type concrete declaration_ hash based ID (string) of the type of the component.
+      - `"typeArguments"`: an array of the _type arguments_ used when applying the type of the _type argument_, if the type is generic, otherwise non existent. The format of the elements of this array recursively follows the rules described in this section.
```

Comment:
```suggestion
```
A metadata type can have components and those components point to types via the `typeId` property. Below is a list of the possible types a component can point to and the state of `typeArguments` based on it:
1. Concrete type - the component won't have `typeArguments` ([concrete type example](https://github.com/FuelLabs/fuel-specs/pull/599/files#diff-29c836429beedee654a72c37d1f883c1e5a0084d468e1446c61da13d7f88783dR405)),
2. Non-generic metadata type - the component won't have `typeArguments` ([non-generic metadata type example](https://github.com/FuelLabs/fuel-specs/pull/599/files#diff-29c836429beedee654a72c37d1f883c1e5a0084d468e1446c61da13d7f88783dR401)),
3. Generic metadata type - the component will have `typeArguments` that point to generics, and those `typeArguments` won't have any `typeArguments` themselves ([generic metadata type example](https://github.com/FuelLabs/fuel-specs/pull/599/files#diff-29c836429beedee654a72c37d1f883c1e5a0084d468e1446c61da13d7f88783dR561))


Based on this, it seems to me that the `typeArguments` property shouldn't exist on a component's type argument anymore. What do you think?